File: windows/tests/ConnectorControl.Core.Tests/ToolRequirementTests.cs
namespace ConnectorControl.Core.Tests;

public class ToolRequirementTests
{
    [Theory]
    [InlineData("npx", Tool.Npx)]
    [InlineData("node", Tool.Node)]
    [InlineData("uvx", Tool.Uvx)]
    [InlineData("uv", Tool.Uv)]
    [InlineData("NPX.CMD", Tool.Npx)]
    [InlineData("node.exe", Tool.Node)]
    [InlineData(" Uvx ", Tool.Uvx)]
    [InlineData("python", null)]
    [InlineData("", null)]
    [InlineData("   ", null)]
    [InlineData("npx.cmd.exe", null)]   // only one suffix is stripped
    public void RecognisesTheFourToolsByBasename(string command, Tool? expected)
    {
        Assert.Equal(expected, ToolRequirement.RequiredTool(command, []));
    }

    [Fact]
    public void UnwrapsOneCmdSlashC()
    {
        Assert.Equal(Tool.Npx, ToolRequirement.RequiredTool("cmd", ["/c", "npx", "-y", "mcp-remote", "https://x.dev/mcp"]));
        Assert.Equal(Tool.Uvx, ToolRequirement.RequiredTool("cmd.exe", ["/C", "uvx"]));
        Assert.Null(ToolRequirement.RequiredTool("cmd", ["/c"]));
        Assert.Null(ToolRequirement.RequiredTool("cmd", ["/k", "npx"]));
        Assert.Null(ToolRequirement.RequiredTool("cmd", ["/c", "cmd", "/c", "npx"]));   // one level only
    }

    [Fact]
    public void LeavesPathsAlone()
    {
        Assert.Null(ToolRequirement.RequiredTool("/usr/local/bin/npx", []));
        Assert.Null(ToolRequirement.RequiredTool(@"C:\Program Files\nodejs\npx.cmd", []));
        Assert.Null(ToolRequirement.RequiredTool("./node", []));
        Assert.Null(ToolRequirement.RequiredTool("cmd", ["/c", "/opt/homebrew/bin/npx"]));
    }

    [Fact]
    public void ConfigOverloadReadsCommandAndArgs()
    {
        Assert.Equal(Tool.Npx, ToolRequirement.RequiredTool(RemotePattern.Make("https://x.dev/mcp", RemoteLaunchStyle.CmdNpx)));
        Assert.Equal(Tool.Npx, ToolRequirement.RequiredTool(RemotePattern.Make("https://x.dev/mcp", RemoteLaunchStyle.Npx)));
        Assert.Equal(Tool.Node, ToolRequirement.RequiredTool(JsonValue.Object(("command", JsonValue.String("node")))));
        Assert.Null(ToolRequirement.RequiredTool(JsonValue.Object(("args", JsonValue.Array([JsonValue.String("npx")])))));
        // a non-string arg empties the args
        Assert.Null(ToolRequirement.RequiredTool(JsonValue.Object(("command", JsonValue.String("cmd")), ("args", JsonValue.Array([JsonValue.String("/c"), JsonValue.Int(42)])))));
        Assert.Null(ToolRequirement.RequiredTool(JsonValue.String("npx")));
    }

    [Fact]
    public void RequiredToolsAcrossConfigsIsDedupedAndOrdered()
    {
        static JsonValue Cmd(string command, params string[] args) => JsonValue.Object(
            ("command", JsonValue.String(command)),
            ("args", JsonValue.Array(args.Select(JsonValue.String))));

        JsonValue[] configs =
        [
            Cmd("npx", "-y", "mcp-remote", "https://a.dev/mcp"),
            Cmd("cmd", "/c", "npx", "-y", "mcp-remote", "https://b.dev/mcp"),   // the same tool, wrapped
            Cmd("uvx", "some-server"),
            Cmd(@"C:\Program Files\nodejs\node.exe"),                            // a path: no tool
            Cmd("python"),                                                       // not one of the four
            JsonValue.String("not an object"),
        ];
        Assert.Equal([Tool.Npx, Tool.Uvx], ToolRequirement.RequiredTools(configs).ToArray());
        Assert.Empty(ToolRequirement.RequiredTools([]));
        Assert.Empty(ToolRequirement.RequiredTools([Cmd("python")]));
        // ToolInfo.All order (npx, node, uvx, uv), not the configs' order:
        Assert.Equal([Tool.Node, Tool.Uv], ToolRequirement.RequiredTools([Cmd("uv"), Cmd("node")]).ToArray());
    }
}

File: Tests/ConnectorControlCoreTests/ToolRequirementTests.swift
import XCTest
@testable import ConnectorControlCore

final class ToolRequirementTests: XCTestCase {
    func testRecognisesTheFourToolsByBasename() {
        XCTAssertEqual(ToolRequirement.requiredTool(command: "npx", args: ["-y", "mcp-remote", "https://x.dev/mcp"]), .npx)
        XCTAssertEqual(ToolRequirement.requiredTool(command: "node", args: ["server.js"]), .node)
        XCTAssertEqual(ToolRequirement.requiredTool(command: "uvx", args: ["mcp-server-fetch"]), .uvx)
        XCTAssertEqual(ToolRequirement.requiredTool(command: "uv", args: ["run", "server.py"]), .uv)
        XCTAssertNil(ToolRequirement.requiredTool(command: "python", args: []))
        XCTAssertNil(ToolRequirement.requiredTool(command: "", args: []))
        XCTAssertNil(ToolRequirement.requiredTool(command: "   ", args: []))
    }

    func testIgnoresCmdAndExeSuffixesAndCase() {
        XCTAssertEqual(ToolRequirement.requiredTool(command: "NPX.CMD", args: []), .npx)
        XCTAssertEqual(ToolRequirement.requiredTool(command: "node.exe", args: []), .node)
        XCTAssertEqual(ToolRequirement.requiredTool(command: " Uvx ", args: []), .uvx)
        XCTAssertNil(ToolRequirement.requiredTool(command: "npx.cmd.exe", args: []),
                     "only one suffix is stripped")
    }

    func testUnwrapsOneCmdSlashC() {
        XCTAssertEqual(ToolRequirement.requiredTool(
            command: "cmd", args: ["/c", "npx", "-y", "mcp-remote", "https://x.dev/mcp"]), .npx)
        XCTAssertEqual(ToolRequirement.requiredTool(command: "cmd.exe", args: ["/C", "uvx"]), .uvx)
        XCTAssertNil(ToolRequirement.requiredTool(command: "cmd", args: ["/c"]))
        XCTAssertNil(ToolRequirement.requiredTool(command: "cmd", args: ["/k", "npx"]))
        XCTAssertNil(ToolRequirement.requiredTool(command: "cmd", args: ["/c", "cmd", "/c", "npx"]),
                     "one level only")
    }

    func testLeavesPathsAlone() {
        XCTAssertNil(ToolRequirement.requiredTool(command: "/usr/local/bin/npx", args: []))
        XCTAssertNil(ToolRequirement.requiredTool(command: "C:\\Program Files\\nodejs\\npx.cmd", args: []))
        XCTAssertNil(ToolRequirement.requiredTool(command: "./node", args: []))
        XCTAssertNil(ToolRequirement.requiredTool(command: "cmd", args: ["/c", "/opt/homebrew/bin/npx"]))
    }

    func testConfigOverloadReadsCommandAndArgs() {
        XCTAssertEqual(ToolRequirement.requiredTool(for: RemotePattern.make(url: "https://x.dev/mcp")), .npx)
        XCTAssertEqual(ToolRequirement.requiredTool(for: .object([
            "command": .string("cmd"), "args": .array([.string("/c"), .string("npx")])])), .npx)
        XCTAssertEqual(ToolRequirement.requiredTool(for: .object(["command": .string("node")])), .node)
        XCTAssertNil(ToolRequirement.requiredTool(for: .object(["args": .array([.string("npx")])])))
        XCTAssertNil(ToolRequirement.requiredTool(for: .object([
            "command": .string("cmd"), "args": .array([.string("/c"), .int(42)])])),
            "a non-string arg empties the args")
        XCTAssertNil(ToolRequirement.requiredTool(for: .string("npx")))
    }

    func testRequiredToolsAcrossConfigsIsDedupedAndOrdered() {
        let configs: [JSONValue] = [
            RemotePattern.make(url: "https://a.dev/mcp"),                                   // npx
            RemotePattern.make(url: "https://b.dev/mcp"),                                   // npx again
            .object(["command": .string("uvx"), "args": .array([.string("some-server")])]),
            .object(["command": .string("/usr/local/bin/node")]),                           // a path: no tool
            .object(["command": .string("python")]),                                        // not one of the four
            .string("not an object"),
        ]
        XCTAssertEqual(ToolRequirement.requiredTools(for: configs), [.npx, .uvx])
        XCTAssertEqual(ToolRequirement.requiredTools(for: []), [])
        XCTAssertEqual(ToolRequirement.requiredTools(for: [.object(["command": .string("python")])]), [])
        XCTAssertEqual(ToolRequirement.requiredTools(for: [
            .object(["command": .string("uv")]),
            .object(["command": .string("node")]),
        ]), [.node, .uv], "Tool.allCases order, not the configs' order")
    }
}
